VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,989 @@
+  that are specific to how two Python versions are compared.
+
+- The comma separator between constraints is a logical "AND" rather than an
+  "OR". The "OR" does not exist in the syntax making some version ranges
+  harder to express, in particular for vulnerabilities that may affect several
+  exact versions or ranges for multiple parallel release branches. Ranges such as
+  "Django 1.2 or later, or Django 2.2 or later or Django 3.2 or later" are
+  difficult to express without an "OR" logic.
+
+
+Why not use Rubygems requirements notation?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+See:
+
+- https://guides.rubygems.org/patterns/#declaring-dependencies
+
+The Rubygems specification suggests but does not enforce using semver. It uses
+operators similar to the ``node-semver`` spec with the different of the "~>"
+aka. pessimistic operator vs. a plain "~" tilde used in node-semver.  This
+operator implies some semver-like versioning, yet gem version are not strictly
+semver. This makes the notation complex to implement and impractical to reuse
+in places that do not use the same Ruby-specific semver-like semantics.
+
+
+Why not use fewer comparators with only =, >= and <?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For instance, the OSV schema adopts a reduced set of only three comparators:
+
+- "=" is implied when used to enumerate vulnerable versions
+- ">=" (greater or equal) is for the version that introduces a vulnerability
+- "<" (lesser) is for the version that fixes a vulnerability
+
+This approach is simpler and works well for most vulnerable ranges but it faces
+limitations when converting from other notations:
+
+- ">" cannot be converted reliably to ">=" unless you know all the versions and
+  these will never change.
+
+- "<=" cannot be converted reliably to "<" unless you know all the versions and
+  these will never change.
+
+- "!=" cannot be converted reliably: there is no ">" comparator to create an
+  unequal equivalent of "><"; and a combo of ">=" and "<" is not equivalent
+  to inequality unless you know all the versions and these will never change.
+
+
+Why not use richer comparators such as tilde, caret and star?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some existing notations such as used with npm, gem, python, or composer
+provide syntactic shorthand such as:
+
+- a "pessimistic operator" using tilde, ~> or =~  as in "~1.3" or "~>1.2.3" 
+- a caret ^ prefix as in "^ 1.2"
+- using a star in a version segment as in "1.2.*"
+- dash-separated ranges as in "1.2 - 1.4"
+- arbitrary string equality such as "===1.2"
+
+Most of these notations can be converted without loss to the ``vers`` notation.
+Furthermore these notations typically assume a well defined version string
+structure specific to their package ecosystem and are not reusable in another
+ecosystem that would not use the exact same version conventions.
```

Comment:
I see the argument, but pessimistic operators are *widely* supported, because they are useful in a large number of circumstances. Pessimistic operators would make *many* version statements *much* simpler and clearer.

I recommend adding a pessimistic operator, and just defining it in terms of the other operators. It's true that it won't be enough in some cases, but that's fine, then the user has to do something else. But let's not avoid handling a very common case.

Please add something like this:

"The pessmistic operator '~>' is a shorthand; ~> version is shorthand for ">=version|^&lt;higher-version". The higher-version is computed by taking version, removing the last "." and all after it, then incrementing the number at the end of what remains (a number must be present). E.g., "~&gt;2.3.0" is equivalent to ">=2.3.0|&lt;2.4".

Reply:
I don't believe this would fit with `vers`s goal of being generalisable/universal to all ecosystems. '~>' as a shorthand for ">=2.3.0|<2.4" would only make sense when the versioning scheme is well defined (e.g. stridt SemVer) -- here the goal is to accurate represent ranges for every possible versioning scheme/ecosystem out there which may have drastically different rules around what "patch version" means.